src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ActivityExtensions.cs

```diff
@@ -0,0 +1,213 @@
+
+            foreach (var resource in resourceToLibraryAndSpans)
+            {
+                var libraryList = new List<OtlpTrace.InstrumentationLibrarySpans>(resource.Value.Count);
+                foreach (var activitySourceEntry in resource.Value)
+                {
+                    var activitySource = activitySourceEntry.Key;
+                    var otlpLibrarySpans = new OtlpTrace.InstrumentationLibrarySpans
+                    {
+                        InstrumentationLibrary = new OtlpCommon.InstrumentationLibrary
+                        {
+                            Name = activitySource.Name, // Name is enforced to not be null, but it can be empty.
+                            Version = activitySource.Version ?? string.Empty, // NRE throw by proto
+                        },
+                    };
+
+                    otlpLibrarySpans.Spans.AddRange(activitySourceEntry.Value);
+                    libraryList.Add(otlpLibrarySpans);
+                }
+
+                var otlpResources = new OtlpResource.Resource();
+                otlpResources.Attributes.AddRange(
+                    resource.Key.Attributes.Select(ToOtlpAttribute));
+
+                var otlpResourceSpans = new OtlpTrace.ResourceSpans
+                {
+                    Resource = otlpResources,
+                };
+                otlpResourceSpans.InstrumentationLibrarySpans.AddRange(libraryList);
+
+                resourceSpansList.Add(otlpResourceSpans);
+            }
+
+            return resourceSpansList;
+        }
+
+        internal static OtlpTrace.Span ToOtlpSpan(this Activity activity)
+        {
+            // protobuf doesn't understand Span<T> yet: https://github.com/protocolbuffers/protobuf/issues/3431
+            Span<byte> traceIdBytes = stackalloc byte[16];
+            Span<byte> spanIdBytes = stackalloc byte[8];
+
+            activity.TraceId.CopyTo(traceIdBytes);
+            activity.SpanId.CopyTo(spanIdBytes);
+
+            var parentSpanIdString = ByteString.Empty;
+            if (activity.ParentSpanId != default)
+            {
+                Span<byte> parentSpanIdBytes = stackalloc byte[8];
+                activity.ParentSpanId.CopyTo(parentSpanIdBytes);
+                parentSpanIdString = ByteString.CopyFrom(parentSpanIdBytes.ToArray());
+            }
+
+            var startTimeUnixNano = activity.StartTimeUtc.ToUnixTimeNanoseconds();
+            var otlpSpan = new OtlpTrace.Span
+            {
+                Name = activity.OperationName, // TODO: map DisplayName to anything?
+
+                Kind = (OtlpTrace.Span.Types.SpanKind)(activity.Kind + 1), // TODO: there is an offset of 1 on the enum.
+
+                TraceId = ByteString.CopyFrom(traceIdBytes.ToArray()),
+                SpanId = ByteString.CopyFrom(spanIdBytes.ToArray()),
+                ParentSpanId = parentSpanIdString,
+
+                // TODO: Status is still pending, need to pursue OTEL spec change.
```

Comment:
Following up on https://github.com/open-telemetry/opentelemetry-specification/issues/306#issuecomment-628961881

Reply:
This is planned to be discussed in OT Spec SIG tomorrow.
https://docs.google.com/document/d/1-bCYkN-DWJq4jw1ybaDZYYmx-WAe6HnwfWbkm8d57v8/edit#heading=h.evt7z82bt2gy